Add ONC/Sun RPC model and NFS system metrics

johnleslie opened the pull request:
## Changes

Add ONC/Sun RPC model and NFS system metrics

## Merge requirement checklist

* [X] [CONTRIBUTING.md](https://github.com/open-telemetry/semantic-conventions/blob/main/CONTRIBUTING.md) guidelines followed.
* [X] Change log entry added, according to the guidelines in [When to add a changelog entry](https://github.com/open-telemetry/semantic-conventions/blob/main/CONTRIBUTING.md#when-to-add-a-changelog-entry).
  * If your PR does not need a change log, start the PR title with `[chore]`
* [N/A] [schema-next.yaml](https://github.com/open-telemetry/semantic-conventions/blob/main/schema-next.yaml) updated with changes to existing conventions.
  * Added to rpc/registry.yaml:rpc.*  but did not change any existing elements

lmolkova commented:
@johnleslie thanks for the contribution!  
JFYI: we're starting RPC stabilization project https://github.com/open-telemetry/community/pull/2684, which is likely to result in significant refactoring of existing RPC conventions.

Please check out the blog post on the [scope and details](https://opentelemetry.io/blog/2025/stabilizing-rpc-conventions/)

braydonk commented on model/system/metrics.yaml:
```diff
@@ -470,3 +470,333 @@ groups:
+    brief: "Reports the count of kernel NFS client network requests handled"
+    note: |
+      Linux: this metric is taken from the Linux kernel's svc_stat.netcnt, which includes both UDP datagrams and TCP segments
+    instrument: gauge
```
OTLP makes a particular distinction between the `Gauge` and `Counter`/`UpDownCounter` instruments:
https://opentelemetry.io/docs/concepts/signals/metrics/#metric-instruments
https://opentelemetry.io/docs/specs/otel/metrics/supplementary-guidelines/#instrument-selection

The main difference is that `UpDownCounter` is to be used for metrics that track an additive measurement. If I'm not mistaken, every `gauge` in this set of metrics can be either a `Counter`/`UpDownCounter` and none fall into the OTLP definition for `Gauge`. I think they can all be changed to one or the other, where `Counter` is used for values that only ever increase and `UpDownCounter` would be for values that can decrease.

Unless there is something unique about this stat from `nfs`, I believe most of these metrics that would be `Counter`s since they appear to be counting ever-increasing values over time. The only metric I see here that strikes me as an `UpDownCounter` would be `nfsd.thread.count`, assuming that this number could be changed up or down during `nfsd` runtime. 

(Worth noting that some backends will treat `UpDownCounter` and `Gauge` as the same thing, though I think most do make the distinction between `Counter` and `Gauge`)

johnleslie replied:
Most of these metrics do only increase (except `metric.nfs.server.thread.count` which is more of an `updowncounter`), however they can reset to 0 upon server restart or host reboot.. would that make them gauge or counter?

braydonk replied:
It will still classify as a `Counter` under that scenario. All metrics in the OTLP data model will come with a `StartTime`, which is the timestamp of the first point observed in the series. For a `Counter` metric, if a value is detected that is not equal/increasing, this is considered a "counter reset" and the `StartTime` of the series is updated to the new value. Collection technologies (including the Collector where these metrics are currently targeting most closely) are built to handle this for `Counter` metrics automatically.

So I think for this PR, each of these metrics can become a `sum` instrument, and most can be `monotonic: true` other than `nfs.server.thread.count`.

braydonk reviewed (commented):
I think the file permission changed on `model/system/metrics.yaml` and I think it was potentially unintentional. Should that be reverted?

johnleslie commented on model/system/metrics.yaml:
```diff
@@ -470,3 +470,333 @@ groups:
     entity_associations:
       - host
+
+  # metric.system.nfs.* metrics
```
- created nfs-metrics.md using your template and `make table-generation attribute-registry-generation` -- looks good, thanks!
- removed obsolete system.nfs metrics

braydonk commented on docs/nfs/nfs-metrics.md:
```diff
@@ -0,0 +1,246 @@
+
+This document describes instruments and attributes for NFS and NFSD metrics. 
+
+<!-- toc -->
```
I think a `make markdown-toc` will fill this section out.

johnleslie replied:
done

braydonk reviewed (commented):
Nice, just about there!

johnleslie commented on .github/CODEOWNERS:
```diff
@@ -151,4 +151,8 @@
+/model/zos                        @open-telemetry/specs-semconv-approvers @open-telemetry/sig-mainframe-approvers
+
+# NFS
+/docs/nfs                         @open-telemetry/specs-semconv-approvers @open-telemetry/sig-mainframe-approvers
```
done

braydonk commented on .chloggen/model-nfs.yaml:
```diff
@@ -0,0 +1,22 @@
+component: nfs
+
+# A brief description of the change. Surround your text with quotes ("") if it needs to start with a backtick (`).
+note: Add rpc.onc.* / rpc.nfs.* model to support new metric.nfs.*/metric.nfsd.*
```
Suggested change to the changelog note to mention the metrics first and some formatting changes since we know it'll be rendered into Markdown.

```suggestion
note: Add `nfs.*`/`nfsd.*` metrics, along with supporting `rpc.onc.*`/`rpc.nfs.*` attributes.
```

braydonk commented on docs/nfs/nfs-metrics.md:
```diff
@@ -0,0 +1,654 @@
+<!-- prettier-ignore-end -->
+<!-- END AUTOGENERATED TEXT -->
+<!-- endsemconv -->
+
```
I think you can add these to the bottom of the document to resolve markdown lint:
```
[DocumentStatus]: https://opentelemetry.io/docs/specs/otel/document-status
[MetricRecommended]: /docs/general/metric-requirement-level.md#recommended
[MetricOptIn]: /docs/general/metric-requirement-level.md#opt-in
```

johnleslie replied:
Done

thompson-tomo commented on model/nfs/metrics.yaml:
```diff
@@ -0,0 +1,330 @@
+groups:
```
Needs further refinement after other feedback.

johnleslie commented on model/rpc/registry.yaml:
```diff
@@ -274,3 +274,30 @@ groups:
+        type: int
+        stability: development
+        brief: "ONC/Sun RPC procedure number"
+      - id: rpc.onc.procedure.name
```
NFSv4 has both operations and methods, so ideally we wouldn't overload rpc.method with both.

NFSv3 has i.e. OPEN as an RPC method
NFSv4 would have COMPOUND as an RPC method, and OPEN and READ as a NFSv4 operations.

thompson-tomo replied:
I think we can describe the meaning of procedure given that it is a totally different name.

johnleslie replied:
Unfortunately in Linux we don't have combined statistics for operations... we only have counters for operations and counters for methods. Any issue leaving as-is?

jsuereth replied:
@thompson-tomo can you outline, more specifically, what use cases you're seeing here for treating RPC + NFS as the same?

Is this something most o11y users would do?  View NFS saturation the same as other RPC services?

I'm not sure I buy this abstraction as being an improvement to general observability and I'm struggling to understand why we would join these two patterns. 

We haven't written up our guidance (AFAIK) on when to provide abstract namespaces and when to divide, but I feel pushing NFS into RPC is one step too far.

cc @open-telemetry/specs-semconv-maintainers for thoughts.

braydonk replied:
> My preference is for nfs telemetry to be in a storage namespace as commented on this PR. That way we can compare nfs storage volumes to alternatives.

I think it's possible for both of our goals to coexist.

We don't currently have a more generic `storage` namespace. I think it could and should exist. Within the @open-telemetry/semconv-system-approvers group though, we purposely strive to match the instrumentation source with the namespace.

In this case, I can envision instrumentation (focusing on Linux for the sake of example) that would get all the mounted filesystems, and use something like [`statvfs(2)`](https://man7.org/linux/man-pages/man2/statvfs.2.html) on each to get information out of each of them. The filesystem type would be presumably an attribute on an Entity or attribute on the Datapoint (presumably the former but I have not deeply thought about design implications of either). The metrics produced by this information would be under `storage.*`.

With the stated goal of comparing storage volumes of different types, in my opinion this accomplishes the goal more simply and directly. Different filesystems will have all kinds of different stats, and there are likely overlaps between them, but rather than trying to bolt different filesystem stats together in some shared abstract `storage.*` namespace metrics, the `storage.*` namespace only contains the types of stats that can be 1:1 directly compared regardless of filesystem type. I think the `storage` namespace should be separately designed with this in mind, rather than trying to force `nfs` conventions within this PR to fit into a mold that we then need to find ways for all filesystem types to fit their stats into.

The conventions within this PR are directly targeted to instrumentation for `nfs` in particular. There could still be separate `storage` instrumentation that would also get its own information about those storage volumes, but I think that's fine and these conventions existing as they do does not lock us out of that future. These metrics exist to reflect the real life patterns of monitoring systems; there are generic APIs and mechanisms to get information about any storage volumes, and there are specific APIs for getting information about `nfs` storage volumes, and there are separate conventions for the important bits of both.

Given this, I feel pretty strongly that we should move ahead with this PR, and design the storage namespace to be generic and reflect the APIs and mechanism used to get general storage volume information.

---

> What I did suggest is based on PR initial formnis why don't we reuse existing attributes for things like procedure & program rather than introducing specific properties when alternatives are pre-existing

Similar to matching metrics to instrumentation sources, attributes and metadata should similarly match existing protocol definitions and the typical ways the information is used. ONC RPC programs have a particularly designed protocol and ways to identify programs and procedures (referring to [`RFC-5531`](https://datatracker.ietf.org/doc/html/rfc5531#page-4) here) that differs from RPC programs that use different protocols. Having these attributes reflect directly what is in the ONC RPC protocol definition lines up with our goals of metrics/attributes reflecting instrumentation.

johnleslie commented on model/nfs/metrics.yaml:
```diff
@@ -0,0 +1,330 @@
+    unit: "{request}"
+    entity_associations:
+      - host
+
+  - id: metric.nfs.client.net.udp.count
+    type: metric
+    metric_name: nfs.client.net.udp.count
+    stability: development
+    brief: "Reports the count of kernel NFS client UDP datagrams handled"
+    note: |
+      Linux: this metric is taken from the Linux kernel's svc_stat.netudpcnt
+    instrument: counter
+    unit: "{datagram}"
+    entity_associations:
+      - host
+
+  - id: metric.nfs.client.net.tcp.count
+    type: metric
+    metric_name: nfs.client.net.tcp.count
+    stability: development
+    brief: "Reports the count of kernel NFS client TCP segments handled"
+    note: |
+      Linux: this metric is taken from the Linux kernel's svc_stat.nettcpcnt
+    instrument: counter
+    unit: "{segment}"
+    entity_associations:
+      - host
+
+  - id: metric.nfs.client.net.tcp.connection.count
+    type: metric
+    metric_name: nfs.client.net.tcp.connection.count
+    stability: development
+    brief: "Reports the count of kernel NFS client TCP connections accepted"
+    note: |
+      Linux: this metric is taken from the Linux kernel's svc_stat.nettcpconn
+    instrument: counter
+    unit: "{connection}"
+    entity_associations:
+      - host
+
```
This metric is the TCP connection count for all NFS clients under a `host`. How could we isolate a network.* metric (none exists currently) to NFS clients?

thompson-tomo replied:
I would have the metrics include the network.protocol.name attribute.

thompson-tomo replied:
Agree nfs would likely be a small fraction but overall I think it would be a useful addition to the existing metrics. Scenario would be analysing metrics generated by a firewall (which is a system) hence the additional attribute would be useful.

johnleslie replied:
As braydonk mentioned they are reported per-nfs-client and per-nfs-server. What's the proposed schema here?

thompson-tomo replied:
Howabout at a minimum

```
metric.rpc.client.connection.count
  attributes:
    rpc.system
    network.transport
    network.type
```

with a similar solution for the server

johnleslie replied:
For these Linux /proc NFS metrics, I can see:

rpc.system=onc_rpc
network.transport=tcp
network.type=??? it's not broken down by IPv4 or IPv6. do we leave this blank?

Also, we need to specify NFS here. how about if we add "rpc.onc.program.name" as an attribute?

thompson-tomo replied:
Yes leave blank if not known, perhaps should be opt-in.

In relation to `rpc.onc.program.name` that seems too specific however `peer.service.name` does seem to fit for the client metric & `service.name` for the server metric.

johnleslie commented on model/nfs/metrics.yaml:
```diff
@@ -0,0 +1,330 @@
+    entity_associations:
+      - host
+
+  - id: metric.nfs.client.rpc.count
+    type: metric
+    metric_name: nfs.client.rpc.count
+    stability: development
+    brief: "Reports the count of kernel NFS client RPCs handled"
+    note: |
+      Linux: this metric is taken from the Linux kernel's svc_stat.rpccnt
+    instrument: counter
+    unit: "{request}"
+    entity_associations:
+      - host
+
+  - id: metric.nfs.client.rpc.retransmit.count
+    type: metric
+    metric_name: nfs.client.rpc.retransmit.count
+    stability: development
+    brief: "Reports the count of kernel NFS client RPC retransmits"
+    note: |
+      Linux: this metric is taken from the Linux kernel's svc_stat.rpcretrans
+    instrument: counter
+    unit: "{retransmit}"
+    entity_associations:
+      - host
+
+  - id: metric.nfs.client.rpc.authrefresh.count
+    type: metric
+    metric_name: nfs.client.rpc.authrefresh.count
+    stability: development
+    brief: "Reports the count of kernel NFS client RPC authentication refreshes"
+    note: |
+      Linux: this metric is taken from the Linux kernel's svc_stat.rpcauthrefresh
+    instrument: counter
+    unit: "{authrefresh}"
+    entity_associations:
+      - host
+
```
How would we isolate an RPC-namespaced metrics to values specific to NFS clients?

thompson-tomo replied:
I would have an attribute such as storage.volume.format.id in the rpc/storage metrics to enable the necessary drill down.

johnleslie replied:
authrefresh is NFS-specific though.. Can you provide an example?

johnleslie replied:
I don't think access_control is quite relevant here. This metric comes from a Linux-specific counter of the number of times a Sun/ONC RPC client has refresh authentication credentials. Other RPC clients (OSes) may have this counter as well.

What do you recommend we do here? Agree it can be more generic than metric.nfs.client.rpc.authrefresh.count, but I think it should be a Sun/ONC RPC specific metric.

braydonk replied:
Do other RPC clients count retransmit or authrefresh? Whenever I look into supporting materials on either one, everything that comes up is NFS related. These metrics are also designed to match up directly with Sun RPC counters from the kernel, the only consumer of which also appears to be NFS. If it isn't really clearly universally applicable, I'd rather not abstract it to the general RPC namespace just for the sake of it.

johnleslie commented on model/rpc/registry.yaml:
```diff
@@ -274,3 +274,30 @@ groups:
+        stability: development
+        brief: "ONC/Sun RPC program name"
+        examples: ['portmapper', 'nfs']
+      - id: rpc.nfs.operation.name
```
hmm, the metrics themselves are already in the NFS namespace i.e. metric.nfs.client.net.udp.count. I think adding a  network.protocol.name="nfs" would be redundant. Am I misunderstanding?

thompson-tomo replied:
In your scenario it would be redundant however the driver for using the storage namespace would be to have a metric such as metric.storage.volume.operation.count which you can filter based on protocol ie nfs, volume, operation.

This would provide a highly reusable solution and if someone wants to see samba/smb metrics it would be the same base as nfs.

thompson-tomo replied:
I would have an attribute ie storage.volume.volume.action/operation which in v1-3 is populated with the procedure value & v4 is populated with the operation value.

In all cases the procedure is set as the `rpc.method.

johnleslie replied:
For NFSv4 we need to report both operations _and_ procedures. My schema above works, but it is network specific, so I'm not sure if it makes sense to have network.protocol.name on metric.storage.volume.operation.count unless it can be left empty without issue.

 . I.e. for a single NFSv4 request of: procedure=COMPOUND, operation=OPEN,operation=READ we'd result in:

id: metric.storage.volume.operation.count=123+1
  - attributes:
    - ref: rpc.method=COMPOUND
    - ref: network.protocol.name=RPC

id: metric.storage.volume.operation.count=456+1
  - attributes:
    - ref: rpc.method=OPEN
    - ref: network.protocol.name=NFSv4

id: metric.storage.volume.operation.count=789+1
  - attributes:
    - ref: rpc.method=READ
    - ref: network.protocol.name=NFSv4

does this work?

thompson-tomo replied:
I was more thinking along the line of for the metric.
```
id: metric.storage.volume.operation.count
  attributes:
    ref: network.protocol.name=NFS
    ref: storage.volume.volume.action/operation= OPEN
```

That way we have a consistent count of operation/actions.

braydonk replied:
Adding `network.protocol.name` to `storage.volume.operation.count` feels a bit weird. Will all other usages of `storage.volume.operation.count` also use that attribute? The network protocol here just feels like a way to add a dimension for NFS's sake rather than something universally applicable.

thompson-tomo replied:
Alternatively we could go with storage.volume.type that way we are contained with storage

jsuereth replied:
agreed with @braydonk here - @thompson-tomo I think where you push for abstraction, we need to understand the value and how that abstraction will be used.

The goal of having a general "storage" set of metrics is laudable, but don't we already have that?  I see this PR adding the *deep* specific NFS metrics we need to monitor NFS directly.  Do the thing we're pushing to abstract actually provide value?   Should we force a "storage" paradigm beyond the notion of disk operations we already have for system metrics?

I think this decision lies with the @open-telemetry/semconv-system-approvers, but I'll add that I'm also unconvinced this abstraction makes sense here.

johnleslie commented on model/rpc/registry.yaml:
```diff
@@ -274,3 +274,30 @@ groups:
+        type: int
+        stability: development
+        brief: "ONC/Sun RPC program number"
+      - id: rpc.onc.program.name
```
re-added onc_rpc to rpc.system

thompson-tomo replied:
So what I will do is add an id attribute as a part of #2405 and that way you can use `peer.service.*` namespace rather than introducing `rpc.onc.program.*` namespace. Note My initial suggestion was to use `rpc.service`

johnleslie replied:
Hmm, I don't see how peer.service.name works for protocols such RPC and NFS, given the brief and examples:

      - ref: peer.service.name
        brief: "The name of the service to which a request is made, as returned by the AWS SDK."
        examples:
          - DynamoDB
          - S3

thompson-tomo replied:
Note this comment is in relation to `rpc.onc.program.number` & `rpc.onc.program.name` which has a name defined as 

> The service.name of the remote service. 
With an example of
> AuthTokenCache

There was no field for number hence adding of the if.

braydonk reviewed (commented):
Given that these metrics as they exist in the PR line up so closely with what `/proc/net/rpc/nfs` and `/proc/net/rpc/nfsd` provide, and as a result the common stat gathering practices like the `nfsstat` command, I'm skeptical about the usefulness of trying to abstract this stuff to generic `rpc` and `storage` namespaces. It looks like we will be cluttering those namespaces with things that would only make sense for `nfs`, and simultaneously making the usage of the conventions for NFS more difficult. I like what is here currently, and it has an [existing implementation PR in the Collector](https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/40134) proving out the schema. I think there would need to be a really convincing reason to uproot this PR as it is.

braydonk reviewed (commented):
To resolve the CI issues, you'll need to run `make registry-generation` and add a `.` to the end of all briefs (this is a new validation rule that didn't exist when the PR was first written but is being caught now).

braydonk commented on .chloggen/model-nfs.yaml:
```diff
@@ -0,0 +1,22 @@
+component: nfs
+
+# A brief description of the change. Surround your text with quotes ("") if it needs to start with a backtick (`).
+note: Add `nfs.*`/`nfsd.*` metrics, along with supporting `rpc.onc.*`/`rpc.nfs.*` attributes.
```
It seems that `nfsd.*` got changed to `nfs.server`, so we can just say `Add nfs.*` here.

johnleslie commented:
> @open-telemetry/semconv-db-approvers can I get a second review/approval on this?

@arminru @trisch-me can I get a second review/approval on this?

johnleslie commented:
> > @open-telemetry/semconv-db-approvers can I get a second review/approval on this?
> 
> @arminru @trisch-me can I get a second review/approval on this?

@trisch-me @dyladan @tedsuo can I get a second review/approval on this please?

braydonk commented:
@open-telemetry/semconv-db-approvers The changes in this PR related to DB are very minor; there's just a couple documents that have the new RPC enum added. Please take a look when you get a chance.

lmolkova reviewed (changes requested):
It seems that metrics being introduced mirror what's already defined in `nfsd_*`, we'd define them quite differently in OTel.  E.g. we don't report each time series as its own metric and tend to merge multiple counts into one while using attributes to break them down (e,g, tcp/udp, error conditions, network direction are attributes, not parts of metric name). Is there a way we can define and report these metrics in 'otel way'?

braydonk commented on model/nfs/metrics.yaml:
```diff
@@ -0,0 +1,229 @@
+    stability: development
+    brief: "Reports the count of kernel NFS server RPCs handled."
+    note: |
+      Linux: this metric is taken from the Linux kernel's svc_stat.rpccnt, the count of good RPCs. this metric can have
```
```suggestion
      Linux: this metric is taken from the Linux kernel's svc_stat.rpccnt, the count of good RPCs. This metric can have
```

johnleslie commented:
> It seems that metrics being introduced mirror what's already defined in `nfsd_*`, we'd define them quite differently in OTel. E.g. we don't report each time series as its own metric and tend to merge multiple counts into one while using attributes to break them down (e,g, tcp/udp, error conditions, network direction are attributes, not parts of metric name). Is there a way we can define and report these metrics in 'otel way'?

Anything else before you can approve? I've addressed all of your feedback. Thanks!

johnleslie commented:
> > > It seems that metrics being introduced mirror what's already defined in `nfsd_*`, we'd define them quite differently in OTel. E.g. we don't report each time series as its own metric and tend to merge multiple counts into one while using attributes to break them down (e,g, tcp/udp, error conditions, network direction are attributes, not parts of metric name). Is there a way we can define and report these metrics in 'otel way'?
> > 
> > 
> > Anything else before you can approve? I've addressed all of your feedback. Thanks!
> 
> Closing

Friendly ping on this. Can you close this request? Trying to get this merged.

lmolkova commented on model/rpc/registry.yaml:
```diff
@@ -274,3 +278,30 @@ groups:
         type: int
         stability: development
         brief: "Uncompressed size of the message in bytes."
+      - id: rpc.onc.version
```
> Once it's onc_rpc, hopefully in a future PR, would it go in a new model/onc_rpc/ dir?

Yes, but  is there a reason not to do it right away and this way minimize churn?

johnleslie replied:
I'd like to get this merged, as it's been several months of review here and it's blocking a migration to OpenTelemetry. Would like to get to this in a future PR.

johnleslie replied:
@lmolkova Any chance you could suggest the change? We now have onc_rpc as a member of rpc.system, but I don't know what you want as far as migrating rpc.onc.{version,procedure.number,procedure.name} under rpc.system.onc_rpc.

johnleslie replied:
@lmolkova see above. "Any chance you could suggest the change? We now have onc_rpc as a member of rpc.system, but I don't know what you want as far as migrating rpc.onc.{version,procedure.number,procedure.name} under rpc.system.onc_rpc."

thompson-tomo replied:
@johnleslie see the code suggestion. Effectively any onc_rpc specific attribute becomes onc_rpc.* ie onc_rpc.version rather than rpc.onc.*. This is what has already been done for Cassandra db for example.

johnleslie replied:
Very much appreciated! Just approved https://github.com/johnleslie/semantic-conventions/pull/1

lmolkova commented on model/nfs/metrics.yaml:
```diff
@@ -0,0 +1,330 @@
+    instrument: counter
+    unit: "{request}"
+    entity_associations:
+      - host
```
not sure I understand, this is a client metric and errors are only reported on the server one. 

If it's not possible to report an error on the client metric, let's include it into the description that this metric only reports successful RPCs

johnleslie replied:
Updated to: Reports the count of kernel NFS client RPCs sent, regardless of whether they're accepted/rejected by the server.

johnleslie replied:
Updated the units to "{request}", and updated the note:

     note: |
       Linux: this metric is taken from the Linux kernel NFSD_STATS_RC_HITS counter in the nfsd_net struct
+      Linux: To get the total number of repcache requests, add hits, misses, and nocache.

lmolkova commented on model/nfs/metrics.yaml:
```diff
@@ -0,0 +1,229 @@
+      - host
+
+  # metric.nfs.server.* metrics
+  - id: metric.nfs.server.repcache.hits
+    type: metric
+    metric_name: nfs.server.repcache.hits
+    stability: development
+    brief: "Reports the kernel NFS server request reply cache hit count."
+    note: |
+      Linux: this metric is taken from the Linux kernel NFSD_STATS_RC_HITS counter in the nfsd_net struct
+    instrument: counter
+    unit: "{hit}"
+    entity_associations:
+      - host
+
+  - id: metric.nfs.server.repcache.misses
+    type: metric
+    metric_name: nfs.server.repcache.misses
+    stability: development
+    brief: "Reports the kernel NFS server request reply cache miss count."
+    note: |
+      Linux: this metric is taken from the Linux kernel NFSD_STATS_RC_MISSES counter in the nfsd_net struct
+    instrument: counter
+    unit: "{miss}"
+    entity_associations:
+      - host
```
if we models a cache hit/miss metric in general, we'd probably have something like `some.cache.operation.count` with attribute `some.cache.response.status = hit | miss`.

I guess the reason to have two metrics is different units, but it makes metrics less discoverable in general and does not make it clear what's the full set of related measurements. In both cases the unit can be changed to `{request}` or  `{response}` and it'd also make it clear that a sum of them is the total number of calls to cache.

johnleslie replied:
Updated the units to "{request}", and updated the note:

     note: |
       Linux: this metric is taken from the Linux kernel NFSD_STATS_RC_HITS counter in the nfsd_net struct
+      Linux: To get the total number of repcache requests, add hits, misses, and nocache.

lmolkova replied:
what I mean is that OTel-idiomatic way would be to define one metric with dimensions. E.g. `nfs.server.cache.requests` with dimension like `nfs.server.cache.state = hit | miss | nocache`. It'd match other metrics and make it easy to consume these metrics. Then explanation on the sum is not even necessary (still helpful)

johnleslie replied:
I think that makes sense. Refactored as metric.nfs.server.repcache.requests with a attribute of nfs.server.repcache.status

lmolkova commented on model/nfs/metrics.yaml:
```diff
@@ -0,0 +1,229 @@
+    entity_associations:
+      - host
+
+  - id: metric.nfs.server.repcache.nocache
```
`nfs.server.repcache.nocache` naming suggests it's one of possible `nfs.server.repcache.*`, but it does nto seem to be, it seems to be related to a response rather than a cache property. IF it's a correct assumption, the suggestion is to call it `nfs.server.reply.nocache` or similar.

johnleslie replied:
This still has to do with the reply cache (repcache). Reworded as:

    brief: "Reports the count of kernel NFS server replies that are uncacheable by the the reply cache."

lmolkova commented on model/nfs/metrics.yaml:
```diff
@@ -0,0 +1,229 @@
+    note: |
+      Linux: this metric is taken from the Linux kernel NFSD_STATS_RC_NOCACHE counter in the nfsd_net struct
+    instrument: counter
+    unit: "{miss}"
```
```suggestion
    unit: "{reply}"
```

?

lmolkova reviewed (commented):
Thanks for addressing most of my feedback, I still have a couple of comments.

johnleslie commented:
@jsuereth @braydonk @lmolkova looks like all comments are resolved -- can you merge?
